src/connector/src/sink/postgres.rs

```diff
@@ -0,0 +1,626 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashSet};
+use std::sync::Arc;
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use itertools::Itertools;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::bitmap::Bitmap;
+use risingwave_common::catalog::Schema;
+use risingwave_common::row::{Row, RowExt};
+use serde_derive::Deserialize;
+use serde_with::{serde_as, DisplayFromStr};
+use simd_json::prelude::ArrayTrait;
+use thiserror_ext::AsReport;
+use tokio_postgres::Statement;
+
+use super::{
+    SinkError, SinkWriterMetrics, SINK_TYPE_APPEND_ONLY, SINK_TYPE_OPTION, SINK_TYPE_UPSERT,
+};
+use crate::connector_common::{create_pg_client, PostgresExternalTable, SslMode};
+use crate::parser::scalar_adapter::{validate_pg_type_to_rw_type, ScalarAdapter};
+use crate::sink::writer::{LogSinkerOf, SinkWriter, SinkWriterExt};
+use crate::sink::{DummySinkCommitCoordinator, Result, Sink, SinkParam, SinkWriterParam};
+
+pub const POSTGRES_SINK: &str = "postgres";
+
+macro_rules! rw_row_to_pg_values {
+    ($row:expr, $statement:expr) => {
+        $row.iter().enumerate().map(|(i, d)| {
+            d.and_then(|d| {
+                let ty = &$statement.params()[i];
+                match ScalarAdapter::from_scalar(d, ty) {
+                    Ok(scalar) => Some(scalar),
+                    Err(e) => {
+                        tracing::error!(error=%e.as_report(), scalar=?d, "Failed to convert scalar to pg value");
+                        None
+                    }
+                }
+            })
+        })
+    };
+}
```

Comment:
I use a `macro` rather than a function call here. The borrow checker does not support returning a reference to a value instantiated within itself, which happens when we call `row.iter()`.